File: backend/modules/petclinic/app/petclinic/api/pettype/getAllPetTypes/GetAllPetTypeService.java
package petclinic.api.pettype.getAllPetTypes;

import akka.actor.ActorRef;
import petclinic.api.Constants;
import petclinic.api.model.PetType;
import petclinic.model.pettypes.PetTypeProtocol;
import scala.collection.JavaConverters;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.Collection;
import java.util.concurrent.CompletionStage;
import java.util.stream.Collectors;

public class GetAllPetTypeService {


    @Inject
    @Named("pettypes")
    private ActorRef petTypesActor;


    public CompletionStage<Collection<PetType>> getAll() {
        return akka.pattern.PatternsCS.ask(petTypesActor, PetTypeProtocol.getAll(), Constants.TIMEOUT)
                .thenApply(obj -> (PetTypeProtocol.GetAllResponse) obj)
                .thenApply(resp -> JavaConverters
                        .asJavaCollection(resp.types())
                        .stream()
                        .map(o -> new PetType(o.id(), o.name()))
                        .collect(Collectors.toList())
                );
    }

}

File: backend/modules/petclinic/app/petclinic/api/pettype/getAllPetTypes/GetAllPetTypeService.scala
package petclinic.api.pettype.getAllPetTypes

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.api.model.PetType
import petclinic.model.pettypes.PetTypeProtocol
import petclinic.model.pettypes.PetTypeProtocol.GetAllResponse

import scala.concurrent.{ExecutionContext, Future}

class GetAllPetTypeService @Inject()(@Named(petclinic.api.Constants.PET_TYPES_ACTOR_NAME) actorRef: ActorRef)(implicit ec: ExecutionContext) {
  def getAll(): Future[Seq[PetType]] = {
    akka.pattern.Patterns.ask(actorRef, PetTypeProtocol.GetAllQuery, Constants.TIMEOUT)
      .mapTo[GetAllResponse]
      .map(x => x.types.map(z => new PetType(z.id, z.name)))
  }

}
